### PR TITLE
New Feature: Event Tree

### PR DESCRIPTION
**Pull request by Herklos**
- Event tree composed by nodes with Event

### CODE REVIEW

File: octobot_commons/event_tree.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+import asyncio
+from asyncio import Event, ALL_COMPLETED
+
+
+class EventTreeNode(object):
+    __slots__ = ['node_value', 'node_event', 'node_type', 'node_task', 'children']
+
+    def __init__(self, node_value, node_type):
+        self.node_value = node_value
+        self.node_type = node_type
```
Comment (GuillaumeDSM) [inline octobot_commons/event_tree.py]: removed node_path 👍

File: octobot_commons/event_tree.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        """
+        self.__set_node(node, value, node_type)
+
+    def set_node_at_path(self, value, node_type, path):
+        """
+        Set the node attributes
+        Creates the node if it doesn't exists
+        :param value: the node 'node_value' attribute to set
+        :param node_type: the node 'node_type' attribute to set
+        :param path: the node path (as a list of string)
```
Comment (GuillaumeDSM) [inline octobot_commons/event_tree.py]:
```
:param path: the node path (as a list of string) => what is the value of these strings ? node_value ?
```

File: octobot_commons/event_tree.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        :return: void
+        """
+        self.__set_node(self.get_node(path), value, node_type)
+
+    def get_node(self, path):
+        """
+        Get the node at the specified path
+        Creates the node if it doesn't exists
+        :param path: the node path (as a list of string)
```
Comment (GuillaumeDSM) [inline octobot_commons/event_tree.py]: same as previous comment :)

File: octobot_commons/event_tree.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+                # update parent node event to gather its children event
+                # TODO think about the total replacement when adding a new node
+                current_node.node_task = asyncio.create_task(self.__set_node_event_from_children(current_node))
+
+                # change to the new node
+                current_node = current_node.children[key]
+
+        return current_node
+
+    def __set_node(self, node, value=None, node_type=None):
```
Comment (GuillaumeDSM) [inline octobot_commons/event_tree.py]: 💯

File: octobot_commons/event_tree.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        Set the node attributes
+        Creates the node if it doesn't exists
+        :param value: the node 'node_value' attribute to set
+        :param node_type: the node 'node_type' attribute to set
+        :param path: the node path (as a list of string)
+        :return: void
+        """
+        self.__set_node(self.get_node(path), value, node_type)
+
+    def get_node(self, path):
```
Comment (GuillaumeDSM) [inline octobot_commons/event_tree.py]: maybe rename to get_or_create_node ?

File: tests/test_event_tree.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        event_tree.set_node(2, None, created_node)
+        asyncio.get_event_loop().call_soon(event_clear.assert_called_once)
+
+
+@pytest.mark.asyncio
+async def test_event_tree_parent_node_event():
+    event_tree = EventTree()
+    created_node = event_tree.get_node(["test"])
+    await asyncio.gather(*[async_event_tree_set_node(event_tree, 2, None, created_node),
+                           event_tree.root.node_event.wait()])
```
Comment (GuillaumeDSM) [inline tests/test_event_tree.py]: 👍

File: octobot_commons/event_tree.py (hunk 5 above)
Comment (Herklos) [reply octobot_commons/event_tree.py]: Good idea :+1:

File: octobot_commons/event_tree.py (hunk 2 above)
Comment (Herklos) [reply octobot_commons/event_tree.py]: No it's a list of keys as string. For example you can pass ["my-parent-1", "my-child-1"] as your node path.

Comment (GuillaumeDSM) [reply octobot_commons/event_tree.py]: I mean can you update the doc with info on what to put in this string to navigate through children ?

Comment (Herklos) [reply octobot_commons/event_tree.py]: Oh yeah ok i'll update the doc

File: octobot_commons/event_tree.py [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+    def set_node_at_path(self, value, node_type, path):
+        """
+        Set the node attributes
+        Creates the node if it doesn't exists
+        :param value: the node 'node_value' attribute to set
+        :param node_type: the node 'node_type' attribute to set
+        :param path: the node path (as a list of string).
+        For example:
+        - If you created a first node with the path ["my-parent-node"]
+        - You can create a child node if my-parent-node by using ["my-parent-node", "my-new-child-node"] as `path`
```
Comment (GuillaumeDSM) [inline octobot_commons/event_tree.py]: I guess it should be child node **in** my-parent-node :)

Comment (Herklos) [reply octobot_commons/event_tree.py]: :+1:

**Pull request merged**